### ISSUE TITLE
Using "zod.infer" in a specific object will cause the vs code typescript intellisense stuck forever in "Loading..."

### ISSUE THREAD

**Issue by ADSNB** [colinhacks/zod#3233]
Hey gyes!

I'm new using zod as my validator for object. I have found a very frustrating bug (I guess) where when I use zod infer to create a new type object based on a zod object my visual studio code intellisense stops working. When this happen I got eternal vs code "Loading..." status and in the bottom corner of the screen says "Loading Intellisense Status..." and this stays forever when I try to use zod infer method.

If we comment z.infer line the intellisense will work:
![Untitled](https://github.com/colinhacks/zod/assets/22732718/f238855a-d5a1-4d47-8a65-7d37c2972986)

If we uncomment z.infer line the intellisense will stop working in entire project:
![Untitled2](https://github.com/colinhacks/zod/assets/22732718/46cbe3f1-e45a-4721-bd37-56ae8afc4eda)

I created a repo where you can reproduce the error with the intellisense. Basically to initialize the repo you should execute "npm install" and "npx prisma generate" to generate the schemas and objects from prisma database.

_**Repo: https://github.com/ADSNB/intellisense-bug-zod-infer**_

I'm very frustrated because we using zod with trpc prisma generator to create class with validation automatically in a big project and this worked for simple tables without relation. But this specific scenario we have a table "ConstructionCompany" and another table "ConstructionCompanyProject" and this tables have an relation one to many. I guess when I used zod infer in this type of object for some reason doesn't work. I tried to use developer console to check the logs and analyse TS server logs but I didn't found anything that could solve this issue.

Any help is really appreciated. If you have any questions let me know.

**Comment by samchungy** [comment]
It's likely that the zod instance your schema was written in may have been created using a different version of zod. The tsc compiler doesn't do too well in that scenario and is likely failing.

**Comment by colinhacks** [comment]
I understand your frustration here and I've encountered some similar infinite loop problems before. 

That's a lot of generated code and there are a lot of recursive types there. Seems that referencing this type in your hook is forcing TypeScript to eagerly resolve an incredibly complex type and it's struggling. Unfortunately I don't have an easy fix right now but it's useful to have a reproduction.

**Comment by medflow-io-luke** [comment]
I have also encountered this error.

I downgraded my zod version and for some reason it started working again but I am not actively calling the z.infer anywhere to my knowledge.

**Comment by jeromeetienne** [comment]
something similar happened to me too. sharing my experience here, i spend a day trying to understand the matter. maybe it will help somebody else 

- it is a large project and i was running multiple Zod from multiple package.json
- it was all the same version (the latest at the time of this writing 3.22.4)
- somehow it became ultra slow

having a single Zod installed in a single package.json made it come back to normal time.

**Comment by aram-wealthmore** [comment]
> something similar happened to me too. sharing my experience here, i spend a day trying to understand the matter. maybe it will help somebody else
> 
> * it is a large project and i was running multiple Zod from multiple package.json
> * it was all the same version (the latest at the time of this writing 3.22.4)
> * somehow it became ultra slow
> 
> having a single Zod installed in a single package.json made it come back to normal time.

Do you mean you were using a monorepo like nx and each app had a zod instance? @jeromeetienne

**Comment by jeromeetienne** [comment]
> Do you mean you were using a monorepo like nx and each app had a zod instance? @jeromeetienne

@aram-wealthmore, yes. i had multiple npm package in a single git repository. (whole code base of a startup)

And vscode was very slow when opening/importing file with Zod. It all solved when i added Zod in the 'top package.json', and removed it to all the 'child package.json'. Make it a single instant of Zod modules

it was 3-4 months ago, and i didnt hit the problem ago. i hope this helps

**Comment by aram-wealthmore** [comment]
> > Do you mean you were using a monorepo like nx and each app had a zod instance? @jeromeetienne
> 
> 
> 
> @aram-wealthmore, yes. i had multiple npm package in a single git repository. (whole code base of a startup)
> 
> 
> 
> And vscode was very slow when opening/importing file with Zod. It all solved when i added Zod in the 'top package.json', and removed it to all the 'child package.json'. Make it a single instant of Zod modules
> 
> 
> 
> it was 3-4 months ago, and i didnt hit the problem ago. i hope this helps

Interesting. I have a similar setup where the expo app in my repo requires its own package json, and everything else pulls from the top level.

However removing didn't seem to resolve the "loading" issue.


This is in a branch I added prisma, zod, and ts-rest in (replacing Joi, pg, and axios)

I only get the loading issue when I am trying to use the types on the front end react app (react native is doing fine)

Going to have to look for a new solution but sharing details here in case anyone has a similar troubleshooting path

**Comment by colinhacks** [comment]
After lots of investigation, I've made some discoveries.

a) This isn't a Zod issue. After a performance improvement in Zod 3.23, this file loads just fine even with the `z.infer` in there. 

b) The problem actually occurs inside React Hook Form, specifically the path generation logic. When you use `<Controller />` RHF infers the type from the `control` prop and uses it to generate dot paths for the `"name"` field. This involves recursing through the entire type which, in this case, is freaking gigantic. Ultimately though, RHF's behavior is behaving as intended and well-implemented (though there's some room for improvement, discussed below) 

c) Fundamentally, this is user error. The type `ConstructionCompanyCreateOneSchema` is absurdly gigantic and recursive. It is not representative of the form you are building, which has three fields. Do something like this to pick the fields you actually care about.

```ts
import {ConstructionCompanyCreateInputObjectSchema} from '@/database/prisma/generated/schemas';
const constructionCompanySchema = ConstructionCompanyCreateInputObjectSchema;

type MyForm = Pick<
  z.infer<typeof schema>,
  'ConstructionCompanyId' | 'Name' | 'CreatedAt'
 >;
```


---

I tweaked RHF's type logic to set a depth limit on key generation. 

With a max depth of 3, there are ~20 keys
![Screenshot 2025-01-13 at 16 06 40](https://github.com/user-attachments/assets/bdcd5ccb-2104-4da5-8279-41fa1d62ee0b)

With a max depth of 4, there are 120ish.
![Screenshot 2025-01-13 at 16 06 32](https://github.com/user-attachments/assets/5340370f-a9bd-4fd8-a5e6-dd0f3a6c07e3)

With a max depth of 5, there are 650ish.
![Screenshot 2025-01-13 at 16 06 23](https://github.com/user-attachments/assets/f962b402-c0e2-4150-92d1-93103c2d354d)

With a max depth of 6 it takes several seconds to load, and there are almost 3400 keys.
![Screenshot 2025-01-13 at 16 08 53](https://github.com/user-attachments/assets/e9f8da93-cab6-4008-bd86-897f66ff8821)

Somewhat ironically, the problem is compounded greatly by RHF's approach to preventing infinite loops, which involves a lot of expensive type comparisons. When I removed al of these guards, TypeScript was able to efficiently produce a list of ~1900 keys. I believe TypeScript aborts the operation as a certain depth level automatically.

![Screenshot 2025-01-13 at 16 20 52](https://github.com/user-attachments/assets/7159ef71-0fcb-45e7-85d2-3e91d0b8dd7b)

But predictably TypeScript also detects the infinite loops and throws an error.

![Screenshot 2025-01-13 at 16 18 57](https://github.com/user-attachments/assets/0a25aeb9-1000-4588-b081-fc7a3c74f14a)

---

Seems to be the best solution to this would be a depth limit in RHF. It's easy enough to implement and if @bluebill1049 is amenable I'd be happy to PR this. It would be a breaking change but it's quite elegant and will speed up TS performance dramatically compared to the current `TraversedTypes` approach.

**Comment by colinhacks** [comment]
Friendly competition is always great of course, but I do want to point out a couple things.

1. The root cause here was the complexity of the type being passed into React Hook Form. It's still valid to use something like `prisma-trpc-generator`, then using `Pick/Omit` to derive the appropriate type for your form.
2. While `prisma-trpc-generator` does generate some very complex types, that's because the goal of that project is to auto-[generate a tRPC router ](https://www.npmjs.com/package/prisma-trpc-generator) that mirrors Prisma's API one-to-one. That involves creating a lot of intermediate types. (You can configure `prisma-trpc-generator` to skip Zod generation.)
3. Other projects exist for generating Zod schemas from Prisma, notably https://github.com/CarterGrimmeisen/zod-prisma which is quite popular.

**Comment by colinhacks** [comment]
Also, if you've read this far, you might be interested in my comment on React Hook Form with some additional thoughts on the `name` path generation problem: https://github.com/react-hook-form/react-hook-form/issues/12167#issuecomment-2588566473
